transport/RtcpReportsProducer.cpp

```diff
@@ -0,0 +1,335 @@
+#include "RtcpReportProducer.h"
+#include "rtp/RtcpFeedback.h"
+#include "rtp/RtcpHeader.h"
+#include <algorithm>
+
+using namespace transport;
+
+namespace
+{
+
+bool ensurePacket(memory::UniquePacket& packet, memory::PacketPoolAllocator& allocator)
+{
+    if (!packet)
+    {
+        packet = memory::makeUniquePacket(allocator);
+    }
+
+    return !!packet;
+}
+
+} // namespace
+
+RtcpReportProducer::RtcpReportProducer(const logger::LoggableId& loggableId,
+    const config::Config& config,
+    const concurrency::MpmcHashmap32<uint32_t, RtpSenderState>& outboundSsrcCounters,
+    const concurrency::MpmcHashmap32<uint32_t, RtpReceiveState>& inboundSsrcCounters,
+    memory::PacketPoolAllocator& rtcpPacketAllocator,
+    RtcpSender& rtcpSender)
+    : _loggableId(loggableId),
+      _config(config),
+      _outboundSsrcCounters(outboundSsrcCounters),
+      _inboundSsrcCounters(inboundSsrcCounters),
+      _rtcpPacketAllocator(rtcpPacketAllocator),
+      _rtcpSender(rtcpSender)
+{
+}
+
+// Send sender reports and receiver reports as needed for the inbound and outbound ssrcs.
+// each sender report send time is offset 1/65536 sec to make them unique when we look them up
+// after receive blocks arrive as they reference the SR by ntp timestamp
+bool RtcpReportProducer::sendReports(uint64_t timestamp, const utils::Optional<uint64_t>& rembMediaBps)
+{
+    uint32_t senderReportSsrcs[_outboundSsrcCounters.capacity()];
+    uint32_t receiverReportSsrcs[_inboundSsrcCounters.capacity()];
+    uint32_t activeSsrcs[_inboundSsrcCounters.capacity()];
+
+    ReportContext reportContext;
+    reportContext.senderReportSsrcs = utils::Span<uint32_t>(senderReportSsrcs, _outboundSsrcCounters.capacity());
+    reportContext.receiverReportSsrcs = utils::Span<uint32_t>(receiverReportSsrcs, _inboundSsrcCounters.capacity());
+    reportContext.activeSsrcs = utils::Span<uint32_t>(activeSsrcs, _inboundSsrcCounters.capacity());
+
+    collectSsrcsReadyToSendSenderReports(reportContext, timestamp);
```

Comment:
inside this method. Why do we track sr count in separate value in ReportContext. Would it not be cleaner to update the Span instead?

Reply:
Yes, I will do that